### PR TITLE
Revert "Merge pull request #155048 from mweinelt/python2-must-die"

### PR DESCRIPTION
**Pull request by Profpatsch**
No. You can’t just delete software people depend on as their daily
driver based on a skewed idea of deprecation.

This reverts commit a97ae545f3a2f74d18affe058db97bfb95990b5a, reversing
changes made to 37d3f7569dfb45182d62849ef5ac871c4a16f462.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Profpatsch** [comment]
For reference, I use lilyterm as my main terminal emulator application, and it didn’t just suddenly stop working because python2 hit an arbitrary milestone.

**Comment by jtojnar** [comment]
Also, I believe @samueldr was working on GTK 3-based lilyterm.

File: pkgs/desktops/gnome-2/default.nix [hunk 1]
```diff
@@ -62,6 +62,12 @@ lib.makeScope pkgs.newScope (self: with self; {
     autoreconfHook = pkgs.autoreconfHook269;
   };
 
+  vte = callPackage ./desktop/vte { };
+
+#### BINDINGS
+
+  libglademm = callPackage ./bindings/libglademm { };
```
[the remark this answers is not in the document]
Comment (Profpatsch) [reply pkgs/desktops/gnome-2/default.nix:69]: This is curious, I reverted the merge commit, maybe something went wrong.

[the remark this answers is not in the document]
Comment (piegamesde) [reply pkgs/desktops/gnome-2/default.nix:69]: I have bad new for you #155061

[the remark this answers is not in the document]
Comment (Profpatsch) [reply pkgs/desktops/gnome-2/default.nix:69]: What happened to deprecation periods?

**Comment by Profpatsch** [comment]
@AndersonTorres I am not okay with removing software I maintain without a deprecation period.

File: pkgs/desktops/gnome-2/default.nix (hunk 1 above)
[the remark this answers is not in the document]
Comment (piegamesde) [reply pkgs/desktops/gnome-2/default.nix:69]: What deprecation periods do you mean? I never heard of any.

**Comment by piegamesde** [comment]
> No. You can’t just delete software people depend on as their daily
driver based on a skewed idea of deprecation.

Python 2 is EOL. Gtk 2 is EOL. This is not an "arbitrary milestone". It means that people stop investing their free time in keeping besaid software secure and running.

Over time, all packages need to be fixed, updated or removed. This is what other distros do as well. A lot of these packages are abandoned, unmaintained – either upstream, in nixpkgs, or both – or have a newer successor. Of course, I prefer it when packages can be fixed and don't need to be removed. I use some of these packages too.

In #157959, I tried to ping all maintainers of affected packages. I am sorry if you somehow weren't noticed in there.

I you can bring a few packages back without the old dependencies (like in #157868 for example), that would be awesome. Otherwise, we are in the fortunate situation that we can simply pin old `nixpkgs` and find the package in there.

**Comment by 7c6f434c** [comment]
We should stop treating the harmful situation that we tolerate software systems needing constant changes as desirable.

**Comment by Profpatsch** [comment]
This throws a very bad light on Python, GTK and also the nix maintainers of said environments. I don’t trust them anymore to support software I use for more than a few years. I will adjust my views accordingly.

For the record, lilyterm does not depend on Python, it depends on vte, which had an optional dependency on Python, so this should have been unnecessary.

**Comment by piegamesde** [comment]
> I don’t trust them anymore to support software I use for more than a few years.

Gtk 2 has been in LTS mode for *over a decade*. Python 2 has prolonged their support frame for several years compared to the planned EOL. For reference, the commercial Java ecosystem wants to get paid after a few years for support.

Just because some software is not supported anymore does not make it instantly go away. The old nixpkgs versions are still there, and they won't change. But please understand that people do not want to invest their free time in keeping some obscure software with a handful of users from a past era alive.

> For the record, lilyterm does not depend on Python, it depends on vte, which had an optional dependency on Python, so this should have been unnecessary.

Maybe, yes. Again, if you can bring back the package without those dependencies that would be awesome.

**Comment by AndersonTorres** [comment]
@Profpatsch 

I have tried to build lilyterm-git. Here is it:

```
lilyterm> unpacking sources
lilyterm> unpacking source archive /nix/store/5mnyj36j878vc5z5fxhdzxjx2k3xbjc5-source
lilyterm> source root is source
lilyterm> patching sources
lilyterm> autoreconfPhase
lilyterm> autoreconf: export WARNINGS=
lilyterm> autoreconf: Entering directory '.'
lilyterm> autoreconf: configure.ac: not using Gettext
lilyterm> autoreconf: running: aclocal --force
lilyterm> configure.ac:16: warning: macro 'AM_PATH_GTK_2_0' not found in library
lilyterm> autoreconf: configure.ac: tracing
lilyterm> autoreconf: configure.ac: not using Libtool
lilyterm> autoreconf: running: intltoolize --copy --force
lilyterm> autoreconf: configure.ac: not using Gtkdoc
lilyterm> autoreconf: running: /nix/store/zmd15qlyb2w92hldhgwavv1682ib3f6z-autoconf-2.71/bin/autoconf --force
lilyterm> configure.ac:5: warning: AC_INIT: not a literal: "$_PACKAGE"
lilyterm> configure.ac:5: warning: AC_INIT: not a literal: "$_VERSION"
lilyterm> configure.ac:5: warning: AC_INIT: not a literal: "$_BUGREPORT"
lilyterm> configure.ac:19: warning: The macro `AC_HELP_STRING' is obsolete.
lilyterm> configure.ac:19: You should run autoupdate.
lilyterm> ./lib/autoconf/general.m4:204: AC_HELP_STRING is expanded from...
lilyterm> configure.ac:19: the top level
lilyterm> configure.ac:23: warning: The macro `AC_HELP_STRING' is obsolete.
lilyterm> configure.ac:23: You should run autoupdate.
lilyterm> ./lib/autoconf/general.m4:204: AC_HELP_STRING is expanded from...
lilyterm> configure.ac:23: the top level
lilyterm> configure.ac:52: warning: The macro `AC_PROG_INTLTOOL' is obsolete.
lilyterm> configure.ac:52: You should run autoupdate.
lilyterm> aclocal.m4:1187: AC_PROG_INTLTOOL is expanded from...
lilyterm> configure.ac:52: the top level
lilyterm> configure.ac:56: warning: The macro `GLIB_GNU_GETTEXT' is obsolete.
lilyterm> configure.ac:56: You should run autoupdate.
lilyterm> aclocal.m4:604: GLIB_GNU_GETTEXT is expanded from...
lilyterm> aclocal.m4:704: AM_GLIB_GNU_GETTEXT is expanded from...
lilyterm> configure.ac:56: the top level
lilyterm> configure.ac:56: warning: The macro `AC_TRY_LINK' is obsolete.
lilyterm> configure.ac:56: You should run autoupdate.
lilyterm> ./lib/autoconf/general.m4:2920: AC_TRY_LINK is expanded from...
lilyterm> lib/m4sugar/m4sh.m4:692: _AS_IF_ELSE is expanded from...
lilyterm> lib/m4sugar/m4sh.m4:699: AS_IF is expanded from...
lilyterm> ./lib/autoconf/general.m4:2249: AC_CACHE_VAL is expanded from...
lilyterm> ./lib/autoconf/general.m4:2270: AC_CACHE_CHECK is expanded from...
lilyterm> aclocal.m4:289: GLIB_LC_MESSAGES is expanded from...
lilyterm> aclocal.m4:604: GLIB_GNU_GETTEXT is expanded from...
lilyterm> aclocal.m4:704: AM_GLIB_GNU_GETTEXT is expanded from...
lilyterm> configure.ac:56: the top level
lilyterm> configure.ac:56: warning: The macro `AC_TRY_LINK' is obsolete.
lilyterm> configure.ac:56: You should run autoupdate.
lilyterm> ./lib/autoconf/general.m4:2920: AC_TRY_LINK is expanded from...
lilyterm> lib/m4sugar/m4sh.m4:692: _AS_IF_ELSE is expanded from...
lilyterm> lib/m4sugar/m4sh.m4:699: AS_IF is expanded from...
lilyterm> ./lib/autoconf/general.m4:2249: AC_CACHE_VAL is expanded from...
lilyterm> ./lib/autoconf/general.m4:2270: AC_CACHE_CHECK is expanded from...
lilyterm> lib/m4sugar/m4sh.m4:699: AS_IF is expanded from...
lilyterm> ./lib/autoconf/headers.m4:89: _AC_CHECK_HEADER_COMPILE is expanded from...
lilyterm> ./lib/autoconf/headers.m4:56: AC_CHECK_HEADER is expanded from...
lilyterm> aclocal.m4:388: GLIB_WITH_NLS is expanded from...
lilyterm> aclocal.m4:604: GLIB_GNU_GETTEXT is expanded from...
lilyterm> aclocal.m4:704: AM_GLIB_GNU_GETTEXT is expanded from...
lilyterm> configure.ac:56: the top level
lilyterm> configure.ac:56: warning: The macro `AC_TRY_LINK' is obsolete.
lilyterm> configure.ac:56: You should run autoupdate.
lilyterm> ./lib/autoconf/general.m4:2920: AC_TRY_LINK is expanded from...
lilyterm> lib/m4sugar/m4sh.m4:699: AS_IF is expanded from...
lilyterm> ./lib/autoconf/headers.m4:89: _AC_CHECK_HEADER_COMPILE is expanded from...
lilyterm> ./lib/autoconf/headers.m4:56: AC_CHECK_HEADER is expanded from...
lilyterm> aclocal.m4:388: GLIB_WITH_NLS is expanded from...
lilyterm> aclocal.m4:604: GLIB_GNU_GETTEXT is expanded from...
lilyterm> aclocal.m4:704: AM_GLIB_GNU_GETTEXT is expanded from...
lilyterm> configure.ac:56: the top level
lilyterm> configure.ac:56: warning: The macro `AC_OUTPUT_COMMANDS' is obsolete.
lilyterm> configure.ac:56: You should run autoupdate.
lilyterm> ./lib/autoconf/status.m4:1025: AC_OUTPUT_COMMANDS is expanded from...
lilyterm> aclocal.m4:388: GLIB_WITH_NLS is expanded from...
lilyterm> aclocal.m4:604: GLIB_GNU_GETTEXT is expanded from...
lilyterm> aclocal.m4:704: AM_GLIB_GNU_GETTEXT is expanded from...
lilyterm> configure.ac:56: the top level
lilyterm> configure.ac:60: warning: The macro `AC_HEADER_STDC' is obsolete.
lilyterm> configure.ac:60: You should run autoupdate.
lilyterm> ./lib/autoconf/headers.m4:704: AC_HEADER_STDC is expanded from...
lilyterm> configure.ac:60: the top level
lilyterm> configure.ac:16: error: possibly undefined macro: AM_PATH_GTK_2_0
lilyterm>       If this token and others are legitimate, please use m4_pattern_allow.
lilyterm>       See the Autoconf documentation.
lilyterm> autoreconf: error: /nix/store/zmd15qlyb2w92hldhgwavv1682ib3f6z-autoconf-2.71/bin/autoconf failed with exit status: 1
error: builder for '/nix/store/hprzv3jwq32ddqfp94hka7q14j3lnjmj-lilyterm-0.9.9.4+date=2019-07-25.drv' failed with exit code 1;
       last 10 log lines:
       > aclocal.m4:704: AM_GLIB_GNU_GETTEXT is expanded from...
       > configure.ac:56: the top level
       > configure.ac:60: warning: The macro `AC_HEADER_STDC' is obsolete.
       > configure.ac:60: You should run autoupdate.
       > ./lib/autoconf/headers.m4:704: AC_HEADER_STDC is expanded from...
       > configure.ac:60: the top level
       > configure.ac:16: error: possibly undefined macro: AM_PATH_GTK_2_0
       >       If this token and others are legitimate, please use m4_pattern_allow.
       >       See the Autoconf documentation.
       > autoreconf: error: /nix/store/zmd15qlyb2w92hldhgwavv1682ib3f6z-autoconf-2.71/bin/autoconf failed with exit status: 1
       For full logs, run 'nix log /nix/store/hprzv3jwq32ddqfp94hka7q14j3lnjmj-lilyterm-0.9.9.4+date=2019-07-25.drv'.
```

It looks like a hard dependency on GTK2.

**Comment by samueldr** [comment]
Just saying, since I was pinged, I won't be using the Nixpkgs-provided lilyterm since *anyway* I was heavily patching it already to remove features.

But the tip-most git revision *did* work with GTK3, albeit it fails at runtime due to other issues. I removed some misfeatures (to me) causing issues. You probably don't want to use [my form](https://github.com/samueldr/lilyterm), but just in case ¯\\\_(ツ)\_/¯

**Comment by samueldr** [comment]
And I have to agree that removing things hapazardly like that is not a good look. No warning, no grace period (even just one full release cycle).

How can I **rely** on Nixpkgs if things *just vanish*? If ~6 months ago a warning was explicitly placed, I would have had time to look and fix what I was using.

Instead, I had to scramble in two evenings to fixup what is might be the most important piece of software in my software stack.

**Comment by AndersonTorres** [comment]
I will open an issue in Discourse and in matrix about this.

File: pkgs/desktops/gnome-2/default.nix (hunk 1 above)
[the remark this answers is not in the document]
Comment (jonringer) [reply pkgs/desktops/gnome-2/default.nix:69]: It's possible in nixpkgs to have `lib.warn` as part of normal evaluations. Ofborg treats these as errors. We could potentially shove the warning into the alias scope, but I haven't seen anyone do that.

**Comment by jonringer** [comment]
> For reference, I use lilyterm as my main terminal emulator application, and it didn’t just suddenly stop working because python2 hit an arbitrary milestone.

Python2 was originally [announced to be EOL](https://www.python.org/dev/peps/pep-0373/) in 2008, and was meant to be EOL in 2016, then extended to 2020.  14 years is a more than generous timelines for ecosystems to adjust and write python2+3 compatible code.

I think it's fairly safe to say that software still dependent on python2 in 2022 has been abandoned.

**Comment by flokli** [comment]
Isn't nixpkgs unstable the place where we should be able to deprecate/remove things?

People who prefer to have some grace period can always stick to the stable channels, and read release notes before upgrading to a new release.

**Comment by samueldr** [comment]
> Isn't nixpkgs unstable the place where we should be able to deprecate/remove things?
>
> People who prefer to have some grace period can always stick to the stable channels, and read release notes before upgrading to a new release.

Announcements prior to removal should be made, in a way that gives **unstable users** time to act. The lack of procedure for correctly handling deprecation is the root cause of the issue. This is not like moving an attribute somewhere else, this is removing a **still working** package, which users may depend on.

It isn't fair to say that in unstable everything goes. Or else we might as well randomize attribute names every few weeks to keep users on their toes.

**Comment by samueldr** [comment]
> I think it's fairly safe to say that software still dependent on python2 in 2022 has been abandoned.

Without a warning, the maintainer of the package was unable to know they had to look at the packaging of their packages for reliance on python2. If there had been a warning/grace period for a full release, maintainers could have taken a look, and in this particular instance, fix the packaging.

**Comment by AndersonTorres** [comment]
> This is not like moving an attribute somewhere else, this is removing a **still working** package, which users may depend on.

I have tried to compile lilyterm without python2 dependency. Nonetheless, It is not working (because GTK2 dependency).
"Still working" does not apply here.

> It isn't fair to say that in unstable everything goes. Or else we might as well randomize attribute names every few weeks to keep users on their toes.

As far as I remember it does not apply because, well, there is a policy to make them have the same name.

> If there had been a warning/grace period for a full release, maintainers could have taken a look, and in this particular instance, fix the packaging.

Technically the maintainers were called; however by any reason Profpatsch was not called or did not notice.

Indeed I removed evilvte by myself, because it was old and unmaintained and with security breaches.

**Comment by samueldr** [comment]
A proper deprecation scheme would warn *users* too. I misspoke when I said maintainers.

We know that in Nixpkgs, a lot of packages are co-maintained by non-maintainer users.

With a warning period that is done through the evaluation, users like me, and I suppose @Profpatsch, would have had a heads-up to look at the issue.

> I have tried to compile lilyterm without python2 dependency. Nonetheless, It is not working (because GTK2 dependency).
> "Still working" does not apply here.

The “**still working**” here means that before the removal, the existing package was working just fine. Removing a package that is obviously broken already is (imo) different than removing one that is currently still working.

**Comment by jonringer** [comment]
It would be nice to have a deprecation mechanism which is compatible with ofborg and hydra.

Perhaps we could have something that is "the opposite" of how aliases work today, in which users have to explicitly opt out of. And for nixpkgs machinery we can just do something like `config.disregardDeprecations = true;` to avoid emitting stuff to stderr for ofborg and hydra.

Then again, the average user will probably do something like, "Why am I getting a python is deprecation error, when I'm not using it?"

**Comment by samueldr** [comment]
> Then again, the average user will probably do something like, "Why am I getting a python is deprecation error, when I'm not using it?"

That's thinking at the trunk, while we're considering leaf packages here.

I agree that if the deprecation starts at a level low enough that is un-actionable, it wouldn't be useful.

In this particular scenario, the leaf packages that *still build* would have to be mentioned in the notice, somehow. For packages that currently are broken the value would be much much less.

The target of removal was `pygtk`. A good message would state something like `%s depends on pygtk, which is slated for removal by 20yy-mm-dd.`. I'm not sure if it's possible to do in some form of automatic manner.

This way "base" users would have a message explaining *which* package is at fault. While it would be generally unactionable for "base" users, they would be given some time to look for alternatives. More involved users would have a warning (ideally for a full release cycle) about the impending change, and some of them, hopefully, would be able to schedule some time to look at the issue.

**Comment by jonringer** [comment]
> The target of removal was `pygtk`. A good message would state something like `%s depends on pygtk, which is slated for removal by 20yy-mm-dd.`. I'm not sure if it's possible to do in some form of automatic manner.

I don't think this could be done generically. From the nix perspective, it's just evaluating thunks. We would need something like `builtins.unsafeGetTopLevelDrv` in order for nix to go back up the stack and inspect which entrypoint eventually included the direct or transitive dependency.

I could also be complete wrong. cc @edolstra

**Comment by AndersonTorres** [comment]
> Then again, the average user will probably do something like, "Why am I getting a python is deprecation error, when I'm not using it?"

I get the `Package cl is deprecated` every single day I launch my Emacs :)

Maybe a message like "python 2 is deprecated" gets hit by an average user, and it goes reported to a superior instance, like the local sysadmin that will investigate this.

About this, it will be useful some sort of backtrace of this warning - IIRC `nix` has something like `why-depends` subcommand.

**Pull request closed without merging**